File: Java/Generics.java
package Java;

import java.util.List;
import java.util.Arrays;
import java.util.function.Function;

public class Generics {
    public static <T> T findBest(List<T> items, Function<T, Double> getValue) {
        if (items == null || items.isEmpty()) {
            throw new IllegalArgumentException("List cannot be empty or null");
        }
        return items.stream()
                    .max((a, b) -> Double.compare(getValue.apply(a), getValue.apply(b)))
                    .get();  // Stream’s max mimics MaxBy
    }

    // Product class
    static class Product {
        private String name;
        private double price;

        public Product(String name, double price) {
            this.name = name;
            this.price = price;
        }

        public double getPrice() { return price; }
        @Override public String toString() { return name + " ($" + price + ")"; }
    }

    // Employee class
    static class Employee {
        private String name;
        private double salary;

        public Employee(String name, double salary) {
            this.name = name;
            this.salary = salary;
        }

        public double getSalary() { return salary; }
        @Override public String toString() { return name + " ($" + salary + ")"; }
    }

    public static void main(String[] args) {
        List<Product> products = Arrays.asList(
            new Product("Laptop", 999.99),
            new Product("Phone", 699.99),
            new Product("Headphones", 149.99)
        );

        Product bestProduct = findBest(products, Product::getPrice);  // Method reference like C# lambda
        System.out.println("Most expensive product: " + bestProduct);  // Most expensive product: Laptop ($999.99)

        List<Employee> employees = Arrays.asList(
            new Employee("Alice", 75000),
            new Employee("Bob", 82000),
            new Employee("Charlie", 68000)
        );

        Employee topEarner = findBest(employees, Employee::getSalary);
        System.out.println("Top earner: " + topEarner);  // Top earner: Bob ($82000)
    }
}







class Container<T> {
    private T item;

    public Container(T item) {
        this.item = item;
    }

    public T getItem() {
        return item;
    }

    public static void main(String[] args) {
        var intContainer = new Container<>(42);      // Holds an Integer
        var strContainer = new Container<>("hello");  // Holds a String

        System.out.println(intContainer.getItem());  // Outputs: 42
        System.out.println(strContainer.getItem());  // Outputs: hello

        // This won’t compile: type mismatch
        // Container<Integer> wrongContainer = new Container<>("not an int");
    }
}

File: C#/Generics.cs
using System;
using System.Collections.Generic;
using System.Linq;

class Program3
{
    static T FindBest<T>(List<T> items, Func<T, double> getValue)
    {
        if (items == null || items.Count == 0)
            throw new ArgumentException("List cannot be empty or null");
        return items.MaxBy(getValue);  // LINQ’s MaxBy is perfect here
    }

    // Example: Products with prices
    class Product
    {
        public string Name { get; }
        public double Price { get; }

        public Product(string name, double price)
        {
            Name = name;
            Price = price;
        }

        public override string ToString() => $"{Name} (${Price})";
    }

    // Example: Employees with salaries
    class Employee
    {
        public string Name { get; }
        public double Salary { get; }

        public Employee(string name, double salary)
        {
            Name = name;
            Salary = salary;
        }

        public override string ToString() => $"{Name} (${Salary})";
    }

    static void Main5()
    {
        var products = new List<Product>
        {
            new Product("Laptop", 999.99),
            new Product("Phone", 699.99),
            new Product("Headphones", 149.99)
        };

        // Usage
        var bestProduct = FindBest(products, p => p.Price);
        Console.WriteLine($"Most expensive product: {bestProduct}");  // Most expensive product: Laptop ($999.99)

        var employees = new List<Employee>
        {
            new Employee("Alice", 75000),
            new Employee("Bob", 82000),
            new Employee("Charlie", 68000)
        };

        var topEarner = FindBest(employees, e => e.Salary);
        Console.WriteLine($"Top earner: {topEarner}");  // Top earner: Bob ($82000)
    }
}









class Container<T>(T item)
{
    private T item = item;

    public T GetItem() => item;
}

class Program6
{
    static void Main()
    {
        var intContainer = new Container<int>(42);
        var strContainer = new Container<string>("hello");

        Console.WriteLine(intContainer.GetItem());  // Outputs: 42
        Console.WriteLine(strContainer.GetItem());  // Outputs: hello

        // This won’t compile: type mismatch
        // var wrongContainer = new Container<int>("not an int");
    }
}